src/glb-redirect/ipt_GLBREDIRECT.c

```diff
@@ -168,11 +170,41 @@ static unsigned int glbredirect_handle_inner_tcp_generic(struct net *net, struct
 	return glbredirect_send_forwarded_skb(net, skb);
 }
 
+/* Build a fake TCP header from the payload of an ICMP Packet Too Big
+ * message. This is required since ICMPv4 only carries the first 64bit
+ * of the TCP / UDP header.
+ */
+static unsigned int fill_tcp_from_icmp(const struct sk_buff *skb, int icmp_payload_ofs, struct tcphdr *th)
+{
+	/* These are reversed from the usual order, since Packet Too Big messages
+	 * contain packets from the return direction of the flow.
+	 */
+	struct {
+		uint16_t dest;
+		uint16_t source;
+	} __attribute__((packed)) _l4p;
+
+	if (skb_header_pointer(skb, icmp_payload_ofs, sizeof(_l4p), &_l4p) == NULL)
+		return 0;
+
+	memset(th, 0, sizeof(*th));
+	th->source = _l4p.source;
+	th->dest = _l4p.dest;
```

Comment:
Since `is_valid_locally` only uses these 2 fields as well, I wonder if it might make sense to acknowledge that more clearly by passing this struct type instead of `struct tcphdr`. Existing full TCP headers on the normal path could be casted to it (not any slower, but clearer in code), which makes it explicit that the function can only use the first 2 fields.

Reply:
I tried that initially, but is_valid_locally also looks at TCP flags and other fields for the SYN cookie check. I'd have to split those into two structures, and make is_valid_locally deal with the "flags" structure being NULL / empty.

Reply:
You're totally right, it uses much more of the data that I recalled at first glance, I agree your solution here makes sense. :+1:

It also looks like, despite being provided with a tcphdr struct, `is_valid_locally` still references into the packet in certain cases around `nf_ct_get_tuplepr`, likely to the ICMP part of the packet. It may be safe to do this because it's a valid packet, alternatively we could avoid that section when the tcphdr is simulated, or potentially https://github.com/github/glb-director/issues/50 might just remove it entirely (I haven't had a chance to dig in to that issue fully yet).

Reply:
It seems like `nf_ct_get_tuplepr` will just return an error if there is not enough data left in the skb. For IPv4 I expect this to always bail out (since ICMP is only guaranteed to contain the first 64 bit), for IPv6 we might find a tuple. The problematic bit is that `nf_ct_get_tuplepr` won't do the reversal of source and dest, so it really would be best to skip the check.

Agreed that #50 would be a nice fix to this :)

Reply:
We don't have a solution for this unfortunately. I'm guessing that #50 won't change in the near term, so I'll have a think how to skip that lookup.